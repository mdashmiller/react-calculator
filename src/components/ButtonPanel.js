import React from 'react'
import PropTypes from 'prop-types'

const ButtonPanel = ({ handleClick, handleKeyDown, calculate }) => 
	// sends values from button to App state
	<div className="wrapper" onKeyDown={handleKeyDown}>
		<div className="top-row">
			<button className="button func" onClick={ () => handleClick('ce') } autoFocus>ce</button>
			<button className="button func" onClick={ () => handleClick('c') }>c</button>
			<button className="button mod" onClick={ () => handleClick('+/-') }>+/-</button>
		</div>
		<div className="button-row">
			<button className="button number" onClick={ () => handleClick('7') }>7</button>
			<button className="button number" onClick={ () => handleClick('8') }>8</button>
			<button className="button number" onClick={ () => handleClick('9') }>9</button>
			<button className="button op" onClick={ () => handleClick('/') }>/</button>
		</div>
		<div className="button-row">
			<button className="button number" onClick={ () => handleClick('4') }>4</button>
			<button className="button number" onClick={ () => handleClick('5') }>5</button>
			<button className="button number" onClick={ () => handleClick('6') }>6</button>
			<button className="button op" onClick={ () => handleClick('x') }>x</button>
		</div>
		<div className="button-row">
			<button className="button number" onClick={ () => handleClick('1') }>1</button>
			<button className="button number" onClick={ () => handleClick('2') }>2</button>
			<button className="button number" onClick={ () => handleClick('3') }>3</button>
			<button className="button op" onClick={ () => handleClick('-') }>-</button>
		</div>
		<div className="button-row">
			<button className="button number" onClick={ () => handleClick('0') }>0</button>
			<button className="button mod" onClick={ () => handleClick('.') }>.</button>
			<button className="button func" id="calc" onClick={calculate}>=</button>
			<button className="button op" onClick={ () => handleClick('+') }>+</button>
		</div>
	</div>	

ButtonPanel.propTypes = {
	handleClick: PropTypes.func.isRequired,
	handleKeyDown: PropTypes.func.isRequired,
	calculate: PropTypes.func.isRequired
}

export default ButtonPanel
