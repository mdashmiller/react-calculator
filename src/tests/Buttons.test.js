import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import ButtonPanel from '../Buttons'

Enzyme.configure({ adapter: new Adapter() })

const props = {
	handleClick: jest.fn(),
	handleKeyDown: jest.fn(),
	calculate: jest.fn()
}

describe('<ButtonPanel />', () => {
	
	let wrapper

	beforeAll(() => {
		wrapper = shallow(<ButtonPanel { ...props } />)
	})

	it('renders without crashing', () => {
		const div = document.createElement('div')
		ReactDOM.render(<ButtonPanel { ...props } />, div)
		ReactDOM.unmountComponentAtNode(div)
	})

	it('has a valid snapshot', () => {
		const component = renderer.create(
			<ButtonPanel { ...props } />
		)
		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders one top row', () => {
		expect(wrapper.find('div.top-row').length).toBe(1)
	})

	it('renders 4 button rows', () => {
		expect(wrapper.find('div.button-row').length).toBe(4)
	})

	it('renders 19 buttons', () => {
		expect(wrapper.find('.button').length).toBe(19)
	})

	it('calls calculate() when #calc is clicked', () => {
		wrapper.find('button#calc').simulate('click')

		expect(props.calculate).toHaveBeenCalled()
	})

	it('calls handleClick() when the buttons are clicked', () => {
		wrapper.find('.button').not('#calc').forEach(node => {
			node.simulate('click')
		})

		expect(props.handleClick).toHaveBeenCalledTimes(18)
	})

})
