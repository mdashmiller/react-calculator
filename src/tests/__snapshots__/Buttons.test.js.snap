// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonPanel /> has a valid snapshot 1`] = `
<div
  className="btn-frame"
  onKeyDown={[MockFunction]}
>
  <div
    className="top-row"
  >
    <button
      autoFocus={true}
      className="button func"
      onClick={[Function]}
    >
      ce
    </button>
    <button
      className="button func"
      onClick={[Function]}
    >
      c
    </button>
    <button
      className="button mod"
      onClick={[Function]}
    >
      +/-
    </button>
  </div>
  <div
    className="button-row"
  >
    <button
      className="button number"
      onClick={[Function]}
    >
      7
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      8
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      9
    </button>
    <button
      className="button op"
      onClick={[Function]}
    >
      /
    </button>
  </div>
  <div
    className="button-row"
  >
    <button
      className="button number"
      onClick={[Function]}
    >
      4
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      5
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      6
    </button>
    <button
      className="button op"
      onClick={[Function]}
    >
      x
    </button>
  </div>
  <div
    className="button-row"
  >
    <button
      className="button number"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="button number"
      onClick={[Function]}
    >
      3
    </button>
    <button
      className="button op"
      onClick={[Function]}
    >
      -
    </button>
  </div>
  <div
    className="button-row"
  >
    <button
      className="button number"
      onClick={[Function]}
    >
      0
    </button>
    <button
      className="button mod"
      onClick={[Function]}
    >
      .
    </button>
    <button
      className="button func"
      id="calc"
      onClick={[MockFunction]}
    >
      =
    </button>
    <button
      className="button op"
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
